### ISSUE TITLE
Cannot return value to a request

### ISSUE THREAD

**Issue by jsantos98** [amqp/rhea#196]
Hello,

I'm quite new with AMQP 1.0, so I may be missing something very basic, so hopefully, someone may be able to help me on the problem I'm having.

I'm trying to implement an endpoint to communicate with others that are following a standard (IPC-CFX) using AMQP as transport.

I've been able to receive messages without much problem, but, when I try to reply to a direct request, I've been failing miserably.
The request is being done using C# AMPQ assembly, specially with ReceiverLink and SenderLink.

On my side (using rhea), I can receive the message in the "message" event, however, I have no idea how to send the reply back

The message looks like something like:

```
Message {
        durable: true,
        ttl: 30000,
        delivery_count: 0,
        delivery_annotations: [Object],
        message_annotations: [Object],
        message_id: 'command-request',
        reply_to: 'test',
        creation_time: 1550790216916,
        application_properties: [Object],
        body: <Buffer 7b 22 4d 65 73 73 61
```

I'm using Azure Service Bus.

Any pointer on how to accomplish this?

Thanks!

**Comment by grs** [comment]
In general you just need to open a sender to the address indicated in the reply_to property of the message, and send the reply over that. (That assumes that the reply_to property was correctly set by the process making the request).

You can also use a so-called 'anonymous' sender, i.e. a sender without an explicit address, and then set the 'to' field on the response to match the reply_to of the request. Not every server supports that, but a lot do. (This is what the request-response example des: https://github.com/amqp/rhea/blob/master/examples/server.js)

**Comment by jsantos98** [comment]
I did try the anonymous sender, and the results was an internal error in Azure..

`private onMessage(context: MessageReceivedContext): void {
        const channel: string = (<any>context.receiver.options).source.address;
        const messageId: string = (context.message.message_id || "").toString();
        const replyTo: string = (context.message.reply_to || context.message.reply_to_group_id || "").toString();
        const messageBody: Buffer = context.message.body;

        const data = this._subscribeChannels.get(channel);
        if (data != null) {
            const callbackArgs: MessageReceived = {
                messageId: messageId,
                messageBody: messageBody,
            };

            data.callbacks.forEach((cb) => {
                cb(callbackArgs);
            });
        }

        if (replyTo !== "") {
            const request = context.message;
            console.log("Received: " + request.body);
            const response: any = {to: replyTo, body: "This is the Reply"};
            if (request.correlation_id) {
                response.correlation_id = request.correlation_id;
            }
            context.connection.send(response);
        }
    }
`
and got this:
`
Received: {"MessageName":"CFX.InformationSystem.UnitValidation.ValidateUnitsRequest","Version":"1.0","TimeStamp":"2019-02-22T14:20:26.2956165+00:00","UniqueID":"fe385aed-16e8-4722-90ea-feec185d4329","Source":"test2","Target":"Connect.IoT","RequestID":"REQUEST-03ea3aa6-e523-4f45-b6a5-33a62dd0112f","MessageBody":{"Validations":["UnitRouteValidation","UnitStatusValidation"],"PrimaryIdentifier":"TESTUNIT1000","Units":[]}}
2019-02-22 14:20:37.425 error: AMQP error (Connect.IoT): An AMQP error occurred (condition='amqp:internal-error').
`

Having a sender basically did't trigger any reply on .net side.

Any other idea?

**Comment by jsantos98** [comment]
Not sure if helps, but the .net part is this (only the send-request part):

```
                        reqConn = factory.CreateAsync(new Address(targetAddress.ToString())).Result;
                        reqSession = new Session(reqConn);
                        Attach recvAttach = new Attach()
                        {
                            Source = new Source() { Address = request.Target },
                            Target = new Target() { Address = CFXHandle }
                        };

                        receiver = new ReceiverLink(reqSession, "request-receiver", recvAttach, null);
                        receiver.Start(300);
                        sender = new SenderLink(reqSession, CFXHandle, request.Target);

                        sender.Send(req);
                        Message resp = receiver.Receive(RequestTimeout.Value);

// **************** the reply should be received here...

                        if (resp != null)
                        {
                            receiver.Accept(resp);
                            response = AmqpUtilities.EnvelopeFromMessage(resp);
                        }
```

**Comment by grs** [comment]
Are you sure the requester sets the reply-to correctly? You could try with an explicit sender link for each reply.

**Comment by jsantos98** [comment]
Yes, I'm sure... changed the log to show the entire message:

Received: {"durable":true,"ttl":30000,"delivery_count":0,"delivery_annotations":{"x-opt-lock-token":{"type":"Buffer","data":[146,254,223,153,127,198,79,241,162,71,226,193,187,137,0,42]}},"message_annotations":{"x-opt-enqueued-time":1550846196502,"x-opt-sequence-number":{"type":"Buffer","data":[0,42,0,0,0,0,0,10]},"x-opt-enqueue-sequence-number":0,"x-opt-partition-id":42,"x-opt-partition-key":"2","x-opt-locked-until":1550846228159},"message_id":"command-request",**"reply_to":"test2"**,"creation_time":1550846195210,"application_properties":{"offset":1},"body":{"type":"Buffer","data":[123,34,77,101,115, ... ]}}

**Comment by grs** [comment]
and test2 is the address the receiver is subscribed to? (i.e. it is not using a server generated temp address)

I don't know whether anonymous senders are supported with Azure Service Bus, so perhaps worth trying an explicit sender link to test2 and see if that works.

**Comment by jsantos98** [comment]
no, test2 is the address of the sender (.net) and it is not a generated address. I have both queues created:

![image](https://user-images.githubusercontent.com/4644017/53249593-5df51e80-36b0-11e9-873e-1abe9b7671b8.png)

I'll try to change the code again to use the explicit sender...

Thanks!

**Comment by grs** [comment]
test2 should be the address of the .net receiver, so that it can receive the reply.

**Comment by jsantos98** [comment]
Just tested with an explicit sender and the request doesn't receive the reply... however, the reply message does go to the queue:
![image](https://user-images.githubusercontent.com/4644017/53250287-0657b280-36b2-11e9-9209-f5544eb07f5a.png)

**Comment by jsantos98** [comment]
Either I'm not configuring the queues correctly in the Azure, or this is not the expected way to reply to a ReceiverLink/SenderLink.

**Comment by jsantos98** [comment]
> test2 should be the address of the .net receiver, so that it can receive the reply.

Yes, it is the address of the receiver (sender of the request, receiver of the reply)

**Comment by grs** [comment]
The requester should have two links, a sender (over which the request is sent) and a receiver on a different address, over which the response will be received. The reply-to in the message should be the address from the latter.

**Comment by jsantos98** [comment]
Still is not working, don't understand what is going on...

I'll try some other approaches and if I'll find anything, I'll let you know.
If in the meantime someones knows any other thin to try, I'll be glad to test it out.

Thanks
